### PR TITLE
impl(generator): set dependencies between discovery types

### PR DESCRIPTION
**Pull request by scotthart**
part of the work for #11190 

Add dependency edges between the type vertices for later use.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12320)

### CODE REVIEW

File: generator/internal/discovery_to_proto.cc [hunk 1]
```diff
@@ -298,7 +298,7 @@ std::set<std::string> FindAllRefValues(nlohmann::json const& json) {
   } else if (json.contains("additionalProperties") || json.contains("items")) {
     fields = json;
   } else {
-    return {};
+    return ref_values;
```
Comment (devbww) [inline generator/internal/discovery_to_proto.cc:301]: I would suggest dropping the `else` altogether, and allowing an empty `fields` to return an empty `ref_values` without a special case.

File: generator/internal/discovery_to_proto.cc [hunk 2]
```diff
@@ -314,6 +314,28 @@ std::set<std::string> FindAllRefValues(nlohmann::json const& json) {
...
 
+Status EstablishTypeDependencies(
+    std::map<std::string, DiscoveryTypeVertex>& types) {
+  for (auto& type : types) {
+    auto const& json = type.second.json();
+    auto ref_values = FindAllRefValues(json);
+    for (auto const& ref : ref_values) {
+      auto ref_iter = types.find(ref);
+      if (ref_iter == types.end()) {
+        return internal::InternalError(
```
Comment (devbww) [inline generator/internal/discovery_to_proto.cc:325]: Is this really an internal error, or a problem with the input data?

File: generator/internal/discovery_to_proto.cc [hunk 3]
```diff
@@ -314,6 +314,28 @@ std::set<std::string> FindAllRefValues(nlohmann::json const& json) {
...
+            "Unknown depended upon type",
+            GCP_ERROR_INFO()
+                .WithMetadata("dependent type", type.first)
+                .WithMetadata("depended upon type", ref));
```
Comment (devbww) [inline generator/internal/discovery_to_proto.cc:329]: I would suggest that the name of the unknown, depended-upon type is more than metadata, and should be included in the message as well (instead?).

File: generator/internal/discovery_type_vertex.cc [hunk 4]
```diff
@@ -465,9 +464,13 @@ StatusOr<std::string> DiscoveryTypeVertex::JsonToProtobufMessage(
...
-  return absl::StrFormat("name: %s; needs: %s; needed_by: %s", name_,
-                         absl::StrJoin(needs_, ","),
-                         absl::StrJoin(needed_by_, ","));
+  auto formatter = [](std::string* out, DiscoveryTypeVertex* t) {
+    out->append(t->name());
+  };
+  return absl::StrFormat(
+      "name: %s; needs_type_name: %s; needed_by_type_name: %s", name_,
+      absl::StrJoin(needs_type_, ",", formatter),
+      absl::StrJoin(needed_by_type_, ",", formatter));
```
Comment (devbww) [inline generator/internal/discovery_type_vertex.cc:473]: That looks like a situation better suited to `StrCat()` than `StrFormat()`.

File: generator/internal/discovery_type_vertex.h [hunk 5]
```diff
@@ -44,16 +44,16 @@ class DiscoveryTypeVertex {
   std::string const& name() const { return name_; }
   std::string const& package_name() const { return package_name_; }
   nlohmann::json const& json() const { return json_; }
+  std::set<DiscoveryTypeVertex*> needs_type() { return needs_type_; }
+  std::set<DiscoveryTypeVertex*> needed_by_type() { return needed_by_type_; }
```
Comment (devbww) [inline generator/internal/discovery_type_vertex.h:48]: Given the look of the other accessors, it seems like these should probably return by reference too, and particularly so given that there is an ownership question about the set members.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12320)* status: 0 of 5 files reviewed, 5 unresolved discussions (waiting on @devbww)

___
*[`generator/internal/discovery_to_proto.cc` line 301 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12320#gh-1284830444:-NbFjC5LE7qm-TLM2H2K:b-6tmwt0) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/341f6ea50a70453d88cd5e5a345009b13519c45a/generator/internal/discovery_to_proto.cc#L301)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

I would suggest dropping the `else` altogether, and allowing an empty `fields` to return an empty `ref_values` without a special case.
</blockquote></details>

`else` removed
___
*[`generator/internal/discovery_to_proto.cc` line 325 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12320#gh-1284836303:-NbFjD_d6tEtG4C3EwW6:bhrsi03) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/341f6ea50a70453d88cd5e5a345009b13519c45a/generator/internal/discovery_to_proto.cc#L325)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Is this really an internal error, or a problem with the input data?
</blockquote></details>

Changed error type
___
*[`generator/internal/discovery_to_proto.cc` line 329 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12320#gh-1284836930:-NbFjQJi6ahRe8dFLIEC:bsyccss) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/341f6ea50a70453d88cd5e5a345009b13519c45a/generator/internal/discovery_to_proto.cc#L329)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

I would suggest that the name of the unknown, depended-upon type is more than metadata, and should be included in the message as well (instead?).
</blockquote></details>

error message enhanced
___
*[`generator/internal/discovery_type_vertex.h` line 48 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12320#gh-1284839717:-NbFjUOO-XHjVB6bveQX:b-6bogez) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/341f6ea50a70453d88cd5e5a345009b13519c45a/generator/internal/discovery_type_vertex.h#L48)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Given the look of the other accessors, it seems like these should probably return by reference too, and particularly so given that there is an ownership question about the set members.
</blockquote></details>

Changed to return a reference to the `std::set`
___
*[`generator/internal/discovery_type_vertex.cc` line 473 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12320#gh-1284837973:-NbFjiM86SEHl8EPoohB:b-7206pd) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/341f6ea50a70453d88cd5e5a345009b13519c45a/generator/internal/discovery_type_vertex.cc#L473)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

That looks like a situation better suited to `StrCat()` than `StrFormat()`.
</blockquote></details>

changed to `StrCat`

**Review by devbww — APPROVED** [review]

**Pull request merged**